flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java

```diff
@@ -0,0 +1,156 @@
+	private static final int MAX_QUEUED_REQUESTS = 1000;
+
+	private final int maxConcurrentCheckpointAttempts;
+	private final Consumer<Long> rescheduleTrigger;
+	private final Clock clock;
+	private final long minPauseBetweenCheckpoints;
+	private final Supplier<Integer> pendingCheckpointsSizeSupplier;
+	private final Object lock;
+	private final PriorityQueue<CheckpointTriggerRequest> triggerRequestQueue = new PriorityQueue<>((r1, r2) -> {
+		if (r1.props.isSavepoint() != r2.props.isSavepoint()) {
+			return r1.props.isSavepoint() ? -1 : 1;
+		} else if (r1.props.forceCheckpoint() != r2.props.forceCheckpoint()) {
+			return r1.props.forceCheckpoint() ? -1 : 1;
+		} else if (r1.isPeriodic != r2.isPeriodic) {
+			return r1.isPeriodic ? 1 : -1;
+		} else {
+			return Long.compare(r1.timestamp, r2.timestamp);
+		}
+	});
+
+	CheckpointRequestDecider(
+			int maxConcurrentCheckpointAttempts,
+			Consumer<Long> rescheduleTrigger,
+			Clock clock,
+			long minPauseBetweenCheckpoints,
+			Supplier<Integer> pendingCheckpointsSizeSupplier,
+			Object lock) {
+		this.maxConcurrentCheckpointAttempts = maxConcurrentCheckpointAttempts;
+		this.rescheduleTrigger = rescheduleTrigger;
+		this.clock = clock;
+		this.minPauseBetweenCheckpoints = minPauseBetweenCheckpoints;
+		this.pendingCheckpointsSizeSupplier = pendingCheckpointsSizeSupplier;
+		this.lock = lock;
+	}
+
+	Optional<CheckpointTriggerRequest> chooseRequestToExecute(CheckpointTriggerRequest newRequest, boolean isTriggering, long lastCompletionMs) {
+		return chooseRequestToExecute(false, () -> newRequest, isTriggering, lastCompletionMs);
+	}
+
+	Optional<CheckpointTriggerRequest> chooseQueuedRequestToExecute(boolean isTriggering, long lastCompletionMs) {
+		return chooseRequestToExecute(true, triggerRequestQueue::peek, isTriggering, lastCompletionMs);
+	}
+
+	/**
+	 * Choose the next {@link CheckpointTriggerRequest request} to execute based on the provided candidate and the
+	 * current state. Acquires a lock and may update the state.
+	 * @return request to execute, if any.
+	 */
+	private Optional<CheckpointTriggerRequest> chooseRequestToExecute(
+			boolean isEnqueued,
+			Supplier<CheckpointTriggerRequest> candidateSupplier,
+			boolean isTriggering,
+			long lastCompletionMs) {
+		synchronized (lock) {
+			final CheckpointTriggerRequest candidate = candidateSupplier.get();
+			if (candidate == null) {
+				return Optional.empty();
+			} else if (isTriggering) {
+				return onCantTriggerNow(candidate, isEnqueued);
+			} else if (pendingCheckpointsSizeSupplier.get() >= maxConcurrentCheckpointAttempts) {
+				return onCantCheckpointNow(candidate, TOO_MANY_CONCURRENT_CHECKPOINTS, () -> { /* not used */ }, isEnqueued);
+			} else if (nextTriggerDelayMillis(lastCompletionMs) > 0) {
+				return onCantCheckpointNow(candidate, MINIMUM_TIME_BETWEEN_CHECKPOINTS, () -> rescheduleTrigger.accept(nextTriggerDelayMillis(lastCompletionMs)), isEnqueued);
+			} else if (candidate != triggerRequestQueue.peek()) {
```

Comment:
`else if (!isEnqueued)` to be consistent with other such checks? Or maybe introduce
```
boolean isAlreadyEnqueued(candidate) {
  return candidate == triggerRequestQueue.peek();
}
```
?

edit: or as I suggested below, drop the case of enqueued `candidate` altogether ?

Reply:
Changed code to always work through the queue.